lib/AST/ArchetypeBuilder.cpp

```diff
@@ -2028,7 +2028,10 @@ Type ArchetypeBuilder::mapTypeIntoContext(Module *M,
   if (!canType->hasTypeParameter())
     return type;
 
-  assert(genericParams && "dependent type in non-generic context");
+  if (!genericParams) {
+    // Failed: dependent type in non-generic context.
+    return Type();
+  }
```

Comment:
I'm not sure if this is the right thing to do, but the alternative is to duplicate these checks before attempting to map the type. There's an argument that this function shouldn't assert on failure conditions at all - in most other cases, the type-checker keeps going even after discovering invalid declarations.